Example of Optimized SMC inner kernel (#49)
blackjax-devs/sampling-book: book/algorithms/TemperedSMCWithOptimizedInnerKernel.md

      format_version: '1.3'
      jupytext_version: 1.15.2
  kernelspec:
    display_name: Python 3 (ipykernel)
    language: python
    name: python3
---

<!-- #region tags=["remove_cell"] -->
# Tuning inner kernel parameters of SMC
<!-- #endregion -->

```python
import time
import arviz as az
import jax
import numpy as np
from jax import numpy as jnp
import matplotlib.pyplot as plt
import pandas as pd
import functools

```

This notebook is a continuation of `Use Tempered SMC to Improve Exploration of MCMC Methods`.
In that notebook, we tried sampling from a multimodal distribution using HMC, NUTS
and SMC with an HMC kernel. Only the latter was able to get samples from both modes of the distribution.
Recall that when setting the HMC parameters

```{python}
hmc_parameters = dict(
    step_size=1e-4, inverse_mass_matrix=inv_mass_matrix, num_integration_steps=1
)
```
these were fixed across all iterations of SMC. The efficiency of an SMC sampler can be improven by
informing the inner kernel parameters using the particles population. We can tune one or many inner 
kernel parameters before mutating the particles in step $i$, using the particles outputted by step $i-1$.

> This notebook ilustrates such tuning using IRMH (Independent Rosenbluth Metropolis Hastings) with a multivariate normal proposal distribution.

Reviewer: ```suggestion
these were fixed across all iterations of SMC. The efficiency of an SMC sampler can be improved by
informing the inner kernel parameters using the particles population. We can tune one or many inner 
kernel parameters before mutating the particles in step $i$, using the particles outputted by step $i-1$.
This notebook illustrates such tuning using IRMH (Independent Rosenbluth Metropolis-Hastings) with a multivariate normal proposal distribution.
```

loglikelihood = lambda x: -V(x)

def density():
    linspace = jnp.linspace(-2, 2, 5000).reshape(-1, 1)
    lambdas = jnp.linspace(0.0, 1.0, 5)
    prior_logvals = jnp.vectorize(prior_log_prob, signature="(d)->()")(linspace)
    potential_vals = jnp.vectorize(V, signature="(d)->()")(linspace)
    log_res = prior_logvals.reshape(1, -1) - jnp.expand_dims(
        lambdas, 1
    ) * potential_vals.reshape(1, -1)

    density = jnp.exp(log_res)
    normalizing_factor = jnp.sum(density, axis=1, keepdims=True) * (
            linspace[1] - linspace[0]
    )
    density /= normalizing_factor
    return density
```

```python
def initial_particles_multivariate_normal(dimensions, key, n_samples):
    return jax.random.multivariate_normal(
        key, jnp.zeros(dimensions), jnp.eye(dimensions) * 2, (n_samples,)
    )
```

## IRMH without tuning

The proposal distribution is normal with fixed parameters across all iterations.

```python
from blackjax import adaptive_tempered_smc
from blackjax.smc import resampling as resampling, solver
from blackjax import irmh
def irmh_experiment(dimensions, target_ess, num_mcmc_steps):

>     mean = jnp.ones(dimensions) * 0.0

Reviewer: why not?

```suggestion
    mean = jnp.zeros(dimensions)
```

```python
def irmh_full_cov_experiment(dimensions, target_ess, num_mcmc_steps):
    def factory(normal_proposal_parameters):
        means, cov = normal_proposal_parameters
        proposal_distribution = lambda key: jax.random.multivariate_normal(key, means, cov)
        def proposal_logdensity_fn(proposal, state):
            return jnp.log(jax.scipy.stats.multivariate_normal.pdf(state.position, mean=means, cov=cov))

        return functools.partial(irmh.build_kernel(),
                                 proposal_distribution=proposal_distribution,
                                 proposal_logdensity_fn=proposal_logdensity_fn)

    def mcmc_parameter_factory(state, info):
        return particles_means(state.particles), particles_covariance_matrix(state.particles)

    kernel_tuned_proposal = inner_kernel_tuning(
        logprior_fn=prior_log_prob,
        loglikelihood_fn=loglikelihood,
        mcmc_factory=factory,
        mcmc_init_fn=irmh.init,
        resampling_fn=resampling.systematic,
        smc_algorithm=adaptive_tempered_smc,
        mcmc_parameters={},
        mcmc_parameter_factory=mcmc_parameter_factory,

>         initial_parameter_value=(jnp.ones(dimensions) * 0, jnp.diag(jnp.ones(dimensions)) * 2),

Reviewer: ```suggestion
        initial_parameter_value=(jnp.zeros(dimensions), jnp.diag(jnp.ones(dimensions)) * 2),
```

Reviewer 2: you can use `jnp.eye` instead of `jnp.diag(jnp.ones(...`
